### PR TITLE
(#8127) grpc: add v.1.42.0

### LINKED ISSUE #8127: [request] grpc/1.42.0
### Package Details
  * Package Name/Version: **grpc/1.42.0**
  * Changelog: **https://github.com/grpc/grpc/releases/tag/v1.42.0**

The above mentioned version is newly released by the upstream project and not yet available as a recipe.

I'll gonna add this version :-) PR follows

### PR DESCRIPTION
**Pull request by miklelappo**
Version update: **grpc/1.42.0**

Also add: 1.41.1
Also add: 1.41.0
Also add: 1.40.0

close #8127 

---

- [ x ] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [ x ] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [ x ] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [ x ] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Review by madebr — APPROVED** [review]

**Comment by miklelappo** [comment]
@ericLemanissierBot, @ericLemanissier, I'm updating protobuf to 3.17.3 which was blocking for other PR too....

File: recipes/grpc/all/conandata.yml [hunk 1]
```diff
@@ -1,4 +1,16 @@
...
+    sha256: "b2f2620c762427bfeeef96a68c1924319f384e877bc0e084487601e4cc6e434c"
+  "1.41.1":
+    url: "https://github.com/grpc/grpc/archive/refs/tags/v1.41.1.tar.gz"
+    sha256: "12a4a6f8c06b96e38f8576ded76d0b79bce13efd7560ed22134c2f433bc496ad"
+  "1.41.0":
+    url: "https://github.com/grpc/grpc/archive/refs/tags/v1.41.0.tar.gz"
+    sha256: "e5fb30aae1fa1cffa4ce00aa0bbfab908c0b899fcf0bbc30e268367d660d8656"
+  "1.40.0":
+    url: "https://github.com/grpc/grpc/archive/refs/tags/v1.40.0.tar.gz"
+    sha256: "13e7c6460cd979726e5b3b129bb01c34532f115883ac696a75eb7f1d6a9765ed"
```
Comment (madebr) [inline recipes/grpc/all/conandata.yml:13]: Are these intermediate versions required by some project?
We usually only add the most recent version, and add intermediate (older) releases only when needed.

Comment (miklelappo) [reply recipes/grpc/all/conandata.yml:13]: Yes, we'd like to have these too if possible from Conan.

Comment (miklelappo) [reply recipes/grpc/all/conandata.yml:13]: @madebr , why is it a problem? Longer CI times?

Comment (madebr) [reply recipes/grpc/all/conandata.yml:13]: Longer CI times only apply when modifying the py recipe. (the CI is smart enough to not rebuild everything when e,g. only conandata.yml changes).
My main concern with blanket addition of versions is the maintainability. Sometimes this requires us to maintain multiple patches.
But perhaps this is not the case with grpc as you were able to re-use the same patch for all versions.

Comment (miklelappo) [reply recipes/grpc/all/conandata.yml:13]: Understood :-) Let's give it a try.. These seem to be pretty compatible and we cover way more use cases and projects.

**Comment by miklelappo** [comment]
@SSE4 it's same problem as in #8084 and #8130
`protoc: Bad CPU type in executable`. We need to fix protobuf for M1 ASAP or disable M1 builds for protobuf and related components completely... Should I try your shim patch in my GRPC PR?

**Comment by miklelappo** [comment]
@ericLemanissier , @madebr does M1 CI allows running X86 executables in "Compatibility mode"? Maybe allowing this would solve this issue too as recommended by Google.

**Comment by AntonRechkov** [comment]
@ericLemanissier , @madebr, @SSE4 Can we by any chance remove M1 from Conan GRPC CI and track M1 compatibility as an issue to do not gate users with new GRPC versions?

**Comment by ericLemanissier** [comment]
yes, you just have to raise an exception in `validate(self)`: https://docs.conan.io/en/latest/mastering/conditional.html#constrain-settings-and-options

**Comment by miklelappo** [comment]
@ericLemanissier I will update patches today

**Comment by miklelappo** [comment]
@ericLemanissier, pipeline is running since yesterday evening. Do you know if there is a real problem with conan pipeline or just the queue is too big? Same for others PR's I pushed...

**Comment by ericLemanissier** [comment]
Please be patient, there was a slow down in CI time. If you see that the number of pending jobs does not go down in few hours, you can close the PR, and reopen it after 15 seconds, so that the build restart. You can also remove unneeded patch versions of the recipe.

**Comment by miklelappo** [comment]
@ericLemanissier, thanks! CI went through now, couldn't expect this could take more than a day :-)

**Comment by AntonRechkov** [comment]
Can we merge this PR?

**Review by ericLemanissier — APPROVED** [review]

**Review by SSE4 — APPROVED** [review]

**Comment by SSE4** [comment]
> Can we merge this PR?

PRs are merged by the bot, as soon as it gains enough approval. more info: https://github.com/conan-io/conan-center-index/blob/master/docs/review_process.md

**Comment by miklelappo** [comment]
@SSE4, right. I guess the ask was more in the direction to review this change as soon as we can, as some project is waiting for it :-) Let's hope not everybody is on holidays yet and @madebr or/and @uilianries can have a look soon.

Thanks for your work guys!

File: recipes/grpc/all/conanfile.py [hunk 2]
```diff
@@ -80,14 +80,18 @@ def validate(self):
...
+
     def source(self):
         tools.get(**self.conan_data["sources"][self.version], destination=self._source_subfolder, strip_root=True)
-        # Fix the protoc search path for cross compiling
-        tools.replace_in_file(os.path.join(self._source_subfolder, "cmake", "protobuf.cmake"),
-                "find_program(_gRPC_PROTOBUF_PROTOC_EXECUTABLE protoc)",
-                "find_program(_gRPC_PROTOBUF_PROTOC_EXECUTABLE protoc PATHS ENV PATH NO_DEFAULT_PATH)"
-        )
-        if tools.Version(self.version) >= "1.39.0":
+        if tools.Version(self.version) < "1.42.0":
```
Comment (uilianries) [inline recipes/grpc/all/conanfile.py:88]: I would prefer moving to build step, but I'm okay for this PR.

**Review by uilianries — APPROVED** [review]
LGTM

File: recipes/grpc/all/conanfile.py (hunk 2 above)
Comment (miklelappo) [reply recipes/grpc/all/conanfile.py:88]: Thanks a lot! Will do this with a next change...

**Comment by anton-danielsson** [comment]
This not only disables M1 but also iOS cross builds.
What's the issue with protobuf? Are all versions affected?
If it's only an issue with protoc I think it should still be allowed when cross compiling.
Then you can still do -pr:b x86-64-profile.txt pr:h arm-profile.txt for example.

**Comment by anton-danielsson** [comment]
Also, just removing this check makes it work again for us on our CI/CD (M1 Macs).

**Comment by miklelappo** [comment]
@anton-danielsson the reasons for this are described in short here: https://github.com/conan-io/conan-center-index/pull/8084
The problem is: currently there is no way that Conan CI can build a new protobuf for M1 platform.

> What's the issue with protobuf? Are all versions affected?
Right. Google never supported protobuf for M1 platform. Protoc binary is delivered by Google in compatibility mode (x86 binary for arm platform), but this one can't be started by Conan CI, so we never can pass the tests

Also for reference please see:
https://github.com/conan-io/conan-center-index/issues/6005

**Comment by anton-danielsson** [comment]
Thanks. Hope it will be fixed soon. For now we need to copy the recipe, revert the "fix" and upload it on our own remote.
Works for now.

**Pull request merged**